Civi/Core/SettingsManager.php

```diff
@@ -328,38 +394,60 @@ public function flush() {
   }
 
   /**
-   * Get a list of critical system defaults.
-   *
-   * The setting system can be modified by extensions, which means that it's not fully available
-   * during bootstrap -- in particular, defaults cannot be loaded. For a very small number of settings,
-   * we must define defaults before the system bootstraps.
-   *
-   * @param string $entity
+   * Load boot settings from environment and/or settings file
    *
-   * @return array
+   * @param string $settingsPath
+   *    Path to the civicrm.settings.php file
    */
-  private static function getSystemDefaults($entity) {
-    $defaults = [];
-    switch ($entity) {
-      case 'domain':
-        $defaults = [
-          'installed' => FALSE,
-          // The default list of components should be kept in sync with "civicrm_extension.sqldata.php".
-          'enable_components' => ['CiviEvent', 'CiviContribute', 'CiviMember', 'CiviMail', 'CiviReport', 'CiviPledge'],
-          'customFileUploadDir' => '[civicrm.files]/custom/',
-          'imageUploadDir' => '[civicrm.files]/persist/contribute/',
-          'uploadDir' => '[civicrm.files]/upload/',
-          'imageUploadURL' => '[civicrm.files]/persist/contribute/',
-          'extensionsDir' => '[civicrm.files]/ext/',
-          'extensionsURL' => '[civicrm.files]/ext/',
-          'ext_max_depth' => \CRM_Extension_System::DEFAULT_MAX_DEPTH,
-          'resourceBase' => '[civicrm.root]/',
-          'userFrameworkResourceURL' => '[civicrm.root]/',
-        ];
-        break;
+  public static function bootSettings($settingsPath) {
+    // during bootstrap we can't use a more persistent cache
+    $bootSettingsCache = \CRM_Utils_Cache::create([
+      'name' => 'bootSettings',
+      'type' => ['ArrayCache'],
+    ]);
+
+    $bootSettingsManager = new self($bootSettingsCache);
+
+    // check for constants we need to define
+    $bootConstants = SettingsMetadata::getMetadata(['is_contact' => 0, 'is_constant' => TRUE], NULL, FALSE, TRUE);
+
+    // if a constant value has been set using an env var, we need
+    // to jump in and define it now so the env var value takes precedence
+    // over any "define" calls in the civicrm.settings.php
+    // (hopefully these use if (!defined(X))))
+    $envSettingsValues = $bootSettingsManager->getEnvSettingValues('domain');
+
+    foreach ($bootConstants as $key => $meta) {
+      $fqn = $meta['global_name'] ?? NULL;
+      $value = $envSettingsValues[$key] ?? NULL;
+      if ($fqn && !defined($fqn) && !is_null($value)) {
+        define($fqn, $value);
+      }
+    }
+
+    if (file_exists($settingsPath)) {
+      if (!defined('CIVICRM_SETTINGS_PATH')) {
+        define('CIVICRM_SETTINGS_PATH', $settingsPath);
+      }
+      $error = include_once $settingsPath;
+      if ($error === FALSE) {
+        echo "Could not load the settings file at: {$settingsFile}\n";
+        exit();
```

Comment:
1. `$settingsPath` vs `$settingsFile`? Copy-paste error?
2. Looking at the places which load `civicrm.settings.php`, there appear to be mixed signals about whether/how to check this output. (*Lightly-edited stream-of-conscious...*)
    *  Survey
        * __Regular/Foreground__: In the regular routing (for Drupal/Backdrop/WordPress), the init function simply says `return FALSE`. 
            * (*Recoverable error; but no obvious trace/exit-code*)
        * __Regular/Foreground__: In the regular routing (for Joomla), it doesn't have any check like this. 
            * (*It simply proceeds; no error*)
        * __Regular/Foreground__: In the regular routing (for Standalone), it didn't have a check like this. (But this PR adds the `echo`+`exit`.)
        * __Extern/Background__: In the `civicrm.config.php` templates (for Drupal/Backdrop/Joomla), it's like the above (`echo`+`exit`).
            * (*Non-recoverable error; no backtraces; misleading exit-code*)
        * __Extern/Background__: In the `civicrm.config.php` template for WordPress, and in `cv`'s original `Bootstrap.php`, it throws an `Exception`.
           * (*Recoverable error; backtraces and exit-codes are better*)
    * Debugging this kind of `exit()` can be tricky, so I favor only as a last-resort. Survey shows... you don't have to have the `exit()`.
    * This guard likely originates circa PHP4 -- before `try`/`catch`/`throw`/`Exception` existed. 
    * My guess is that `civicrm.config.php.XXXX` did this because they had un-guarded calls to `include_once` -- and this pattern provided a way to give a prettier error-message.  But in this context, you have the `file_exists()` guard (L428). So I don't think you'll ever get a meaningful error-signal on this.
        * (*In theory, you might say: "Well, what if `civicrm.settings.php` has a syntax error or a bad method-call?" That could happen... but this pattern here won't catch that error...*)

__tldr__: Aaaah! `exit()`!!! Aah!

Reply:
I took this from the Drupal `civicrm.config.php` because I thought that was somewhat canonical, but will remove if it is intended for PHP4!

Reply:
(FWIW my assumed rationale was "dont leak anything to web if civicrm.settings.php has a syntax error or a bad method-call"?)

Reply:
Have removed.